rllib/algorithms/algorithm_config.py

```diff
@@ -3909,6 +3935,54 @@ def _validate_to_be_deprecated_settings(self):
                     f"config.framework({self.framework_str})!"
                 )
 
+    def _validate_custom_async_evaluation_function(self, func: Callable):
+        """Checks if the custom async evaluation function conforms to standards.
+
+        First, it is checked, if the passed in element is indeed a function.
+        Then it is checked, if the signature contains the required number of
+        arguments, i.e. `algorithm`, `eval_workers`, `weights_ref`, and
+        `weights_seq_no`.
+        Finally, the source code of the element is checked for the usage of
+        `foreach_worker_async`.
+
+        All of these checks are only intended to guide the user when passing
+        in a custom function to evaluate asynchronously.
+
+        Args:
+            func: A callable passed into the configuration argument
+                `custom_async_evaluation_function`.
+
+        Raises:
+            `ValueError` if the callable is not a function, does not contain
+                three arguments and does not make use of
+                `eval_workers.foreach_worker_async`.
+        """
+
+        # Import the inspect module.
+        from inspect import getsourcelines, isfunction, signature
+
+        # Check, if we have indeed a function.
+        if not isfunction(func):
+            raise ValueError("`custom_async_evaluation_function must be a function.")
+        # Check, if the signature is correct, i.e. three arguments:
+        #   eval_workers, weights_ref, weights_seq_no
+        func_signature = signature(func)
+        if len(func_signature.parameters) != 4:
+            raise ValueError(
+                "`custom_async_eval_function` expects a callable with four "
+                "arguments, namely `algorithm`, `eval_workers`, `weights_ref`, "
+                f"and `weights_seq_no`, but received signature {func_signature}."
+            )
+        # Check, if `foreach_worker_async` is indeed used inside of the
+        # custom evaluation routine.
+        lines, _ = getsourcelines(func)
+        if not any(["foreach_worker_async" in line for line in lines[0]]):
```

Comment:
A little tricky to check for these syntax, but I guess we do need to force the user to use this API.

Reply:
Yes indeed, and i think - not sure yet, but interested in your opinion - we might need even a test for checking, if the user passes in a partial that has the `weights_ref` in it. Otherwise I guess the user will run into the same memory leak you fixed last week.

Reply:
@sven1977 By the way, during coding this PR I ran over the code of the async evaluation with envrunner and what I was wondering about is this line:

```python
_env_runner_num_episodes=(1 if unit == "episodes" else None),
```

Why not 
```python
_env_runner_num_episodes=(eval_config.num_envs_per_worker if unit == "episodes" else None),
```
? Isn't it a waste of resources, if we use only a single episode from the ones sampled in a single run by a worker?

Reply:
Actually, the PR last week did not fix the leak. We are still not sure what it is exactly and still cannot reproduce this on our end (only customer can with certain configs). We'll keep digging though :)

Reply:
Yeah, you are right about the 1 -> `eval_config.num_envs_per_worker`.
Let's fix it.

Reply:
Fixed.